plugins/openssl/openssl.cpp

```diff
@@ -0,0 +1,932 @@
+int passphraseCB(char *passPhraseBuf, int passPhraseBufSize, int rwflag, void *pPassPhraseMB)
+{
+    size32_t len = ((MemoryBuffer*)pPassPhraseMB)->length();
+    if (passPhraseBufSize >= (int)len)
+    {
+        memcpy(passPhraseBuf, ((MemoryBuffer*)pPassPhraseMB)->bufferBase(), len);
+        return len;
+    }
+    PROGLOG("Private Key Passphrase too long (%d bytes), max %d", len, passPhraseBufSize);
+    return 0;
+}
+
+static constexpr int OPENSSL_MAX_CACHE_SIZE = 10;
+static constexpr bool PRINT_STATS = false;
+template <typename T>
+class OpenSSLCache
+{
+public:
+    const T * checkCache(const char * algorithm_name)
+    {
+        for (auto& c : cache)
+        {
+            if (std::get<0>(c) == algorithm_name)
+            {
+                hits++;
+                return std::get<1>(c);
+            }
+        }
+        misses++;
+        const T * newObj = getObjectByName(algorithm_name);
+        if (newObj)
+        {
+            cache.emplace_front(algorithm_name, newObj);
+            if (cache.size() > OPENSSL_MAX_CACHE_SIZE)
+                cache.pop_back();
+        }
+        else
+            failOpenSSLError("adding new object to cache");
+
+        return newObj;
+    };
+
+    void printStatistics() {DBGLOG("OPENSSL %s CACHE STATS: HITS = %d, MISSES = %d", cacheName.c_str(), hits, misses);};
+
+    void init()
+    {
+        setCacheName();
+        hits = 0;
+        misses = 0;
+    };
+
+private:
+    int hits;
+    int misses;
+    std::string cacheName;
+    std::list<std::tuple<std::string, const T *>> cache;
+
+    void setCacheName();
+    const T * getObjectByName(const char * name);
+};
+
+template <>
+void OpenSSLCache<EVP_CIPHER>::setCacheName() { cacheName = "CIPHER"; }
+
+template <>
+void OpenSSLCache<EVP_MD>::setCacheName() { cacheName = "DIGEST"; }
+
+template <>
+const EVP_CIPHER * OpenSSLCache<EVP_CIPHER>::getObjectByName(const char * name) { return EVP_get_cipherbyname(name); }
+
+template <>
+const EVP_MD * OpenSSLCache<EVP_MD>::getObjectByName(const char * name) { return EVP_get_digestbyname(name); }
+
+// PEM Public/Private keys require parsing from a string
+// Store the hash of the original string and parsed key
+class PKeyCache
+{
+public:
+    void init()
+    {
+        hits = 0;
+        misses = 0;
+    };
+
+    EVP_PKEY * checkCache(size32_t keyLen, const char * key, size32_t passphraseLen, const void * passphrase)
+    {
+        for (auto& c : cache)
+        {
+            if (hashc(reinterpret_cast<const byte *>(passphrase), passphraseLen, hashc(reinterpret_cast<const byte *>(key), keyLen, 0)) == std::get<0>(c))
+            {
+                hits++;
+                return std::get<1>(c);
+            }
+        }
+
+        misses++;
+
+        BIO * bio = BIO_new_mem_buf(key, keyLen);
+        if (!bio)
+            failOpenSSLError("creating buffer for EVP_PKEY");
+
+        EVP_PKEY * pkey;
+        if (startsWith(key, "-----BEGIN RSA PUBLIC KEY-----"))
```

Comment:
This public key may not be RSA; it could be ECC. I think you should check for 'contains "PUBLIC KEY-----"' (not sure about the dashes).